core/context.go

```diff
@@ -0,0 +1,260 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+// #include <pwd.h>
+// #include <grp.h>
+// #include <unistd.h>
+import "C"
+
+import (
+	"fmt"
+	log "github.com/DataDrake/waterlog"
+	"os/exec"
+	gouser "os/user"
+	"strconv"
+)
+
+const minimumUID = 1000
+const wheelGroup = "sudo"
+
+// Context contains contextual system data, such as groups, users, and active shells, along with paths to certain binaries
+type Context struct {
+	usermod  string
+	groupadd string
+	groupmod string
+	users    []User
+	groups   []gouser.Group
+	shells   []string
+}
+
+// User is a convenience struct with information on a user entry in /etc/passwd
+type User struct {
+	Name     string
+	Groups   []string
+	IsActive bool
+	IsRoot   bool
+	IsAdmin  bool
+}
+
+// NewContext creates an initialized instance of a Context object
+func NewContext() (*Context, error) {
+	return (&Context{}).init()
+}
+
+// FilterUsers returns users that fit into one or more of the given filters
+func (c *Context) FilterUsers(filters ...string) []User {
+	var filtered = make([]User, 0)
+
+	for _, it := range c.users {
+		switch {
+		case contains(filters, "all"):
+			fallthrough
+		case contains(filters, "active") && it.IsActive:
+			fallthrough
+		case contains(filters, "system") && !it.IsActive:
+			fallthrough
+		case contains(filters, "admin") && (it.IsRoot || it.IsAdmin):
+			filtered = append(filtered, it)
+		}
+	}
+
+	return filtered
+}
+
+// AddToGroup adds a User to a preexisting group
+func (c *Context) AddToGroup(user User, group string) (bool, error) {
+	if !contains(user.Groups, group) {
+		var cmd = &exec.Cmd{
+			Path: c.usermod,
+			Args: []string{c.usermod, "-aG", group, user.Name},
+		}
+
+		if err := cmd.Run(); err != nil {
+			return false, err
+		}
+
+		user.Groups = append(user.Groups, group)
+		return true, nil
+	}
+	return false, nil
+}
+
+// CreateGroup creates a new group with the given name and ID
+func (c *Context) CreateGroup(name string, id string) error {
+	var cmd = &exec.Cmd{
+		Path: c.groupadd,
+		Args: []string{c.groupadd, "-g", id, name},
+	}
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	c.groups = append(c.groups, gouser.Group{
+		Name: name,
+		Gid:  id,
+	})
+
+	return nil
+}
+
+// UpdateGroupID finds a group with the given name and changes its ID to the given ID
+func (c *Context) UpdateGroupID(name string, id string) error {
+	var cmd = &exec.Cmd{
+		Path: c.groupmod,
+		Args: []string{c.groupmod, "-g", id, name},
+	}
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+
+	for _, it := range c.groups {
+		if it.Name == name {
+			it.Gid = id
+			break
+		}
+	}
+
+	return nil
+}
+
+func (c *Context) init() (*Context, error) {
```

Comment:
I don't really think we need the `init()` function (as described earlier), but when you have multiple return values, it is generally customary to use names to avoid `var` declaration lines.